### PR TITLE
Fixed: Various HealthCheck Improvements

### PR DESCRIPTION
**Pull request by bakerboy448**
#### Database Migration
NO

#### Description
- Align Radarr and Sonarr HealthChecks
- Backport various Radarr changes
- Add ProviderAddedEvent
- Cleanup usings
- Add HealthCheck Wiki Fragments
- Add filterBlockedClients to Download Clients

based partially on Radarr cae4faae61fe47ee6dacf334d8c4e4d483f5e5c8

#### Todos
- [] Tests
- [] Wiki Updates

#### Issues Fixed or Closed by this PR

*

### CODE REVIEW

File: src/NzbDrone.Core/Download/DownloadClientProvider.cs [hunk 1]
```diff
@@ -91,9 +91,39 @@ public IEnumerable<IDownloadClient> GetDownloadClients()
             return _downloadClientFactory.GetAvailableProviders();
         }
 
+        public IEnumerable<IDownloadClient> GetDownloadClients(bool filterBlockedClients = false)
```
Comment (markus101) [inline src/NzbDrone.Core/Download/DownloadClientProvider.cs:94]: Shouldn't the now unused `GetDownloadClients()` be removed?

File: src/NzbDrone.Core/Download/DownloadClientProvider.cs [hunk 2]
```diff
@@ -91,9 +91,39 @@ public IEnumerable<IDownloadClient> GetDownloadClients()
...
+        {
+            var blockedClients = _downloadClientStatusService.GetBlockedProviders().ToDictionary(v => v.ProviderId, v => v);
+
+            foreach (var client in clients)
+            {
+                DownloadClientStatus blockedClientStatus;
+
+                if (blockedClients.TryGetValue(client.Definition.Id, out blockedClientStatus))
+                {
+                    _logger.Debug("Temporarily ignoring client {0} till {1} due to recent failures.", client.Definition.Name, blockedClientStatus.DisabledTill.Value.ToLocalTime());
```
Comment (markus101) [inline src/NzbDrone.Core/Download/DownloadClientProvider.cs:121]: `till` -> `until`

File: src/NzbDrone.Core/Download/DownloadClientProvider.cs [hunk 3]
```diff
@@ -91,9 +91,39 @@ public IEnumerable<IDownloadClient> GetDownloadClients()
...
+
+        private IEnumerable<IDownloadClient> FilterBlockedIndexers(IEnumerable<IDownloadClient> clients)
+        {
+            var blockedClients = _downloadClientStatusService.GetBlockedProviders().ToDictionary(v => v.ProviderId, v => v);
+
+            foreach (var client in clients)
+            {
+                DownloadClientStatus blockedClientStatus;
+
+                if (blockedClients.TryGetValue(client.Definition.Id, out blockedClientStatus))
```
Comment (markus101) [inline src/NzbDrone.Core/Download/DownloadClientProvider.cs:119]: `out var blockedClientStatus` and you don't need to predefine it.

File: src/NzbDrone.Core/HealthCheck/Checks/ImportListRootFolderCheck.cs [hunk 4]
```diff
@@ -40,7 +41,7 @@ public override HealthCheck Check()
                     continue;
                 }
 
-                if (!_diskProvider.FolderExists(rootFolderPath))
+                if (rootFolderPath.IsNullOrWhiteSpace() || !_diskProvider.FolderExists(rootFolderPath))
```
Comment (markus101) [inline src/NzbDrone.Core/HealthCheck/Checks/ImportListRootFolderCheck.cs:44]: How would `rootFolderPath` be null or empty? Are we missing validation elsewhere?

If it's null or empty then we're going to add an empty key to the dictionary and tell the user the root folder is missing, instead of telling them it wasn't set.

**Review by markus101 — CHANGES_REQUESTED** [review]

File: src/NzbDrone.Core/HealthCheck/Checks/DownloadClientSortingCheck.cs [hunk 5]
```diff
@@ -1,5 +1,4 @@
 using System;
-using System.Linq;
```
Comment (markus101) [inline src/NzbDrone.Core/HealthCheck/Checks/DownloadClientSortingCheck.cs:2]: All these using clean ups should be part of separate commit, possibly even a separate PR.

File: src/NzbDrone.Core/HealthCheck/Checks/IndexerSearchCheck.cs [hunk 6]
```diff
@@ -23,21 +23,21 @@ public override HealthCheck Check()
 
             if (automaticSearchEnabled.Empty())
             {
-                return new HealthCheck(GetType(), HealthCheckResult.Warning, "No indexers available with Automatic Search enabled, Sonarr will not provide any automatic search results");
+                return new HealthCheck(GetType(), HealthCheckResult.Warning, "No indexers available with Automatic Search enabled, Sonarr will not provide any automatic search results", "#no-indexers-available-with-automatic-search-enabled-sonarr-will-not-provide-any-automatic-search-results");
```
Comment (markus101) [inline src/NzbDrone.Core/HealthCheck/Checks/IndexerSearchCheck.cs:26]: Fragment changes should be in a separate commit as well

**Review by markus101 — COMMENTED** [review]
It's hard to see what is part of the original Radarr commit and what is part of "other cleanup" because this was all done in a single commit.

**Comment by bakerboy448** [comment]
> It's hard to see what is part of the original Radarr commit and what is part of "other cleanup" because this was all done in a single commit.

nothing was cherry picked so the original commit is reference only as mentioned in the PR `based on`
There are too many conflicts and too many commits to try to track down and pull 1x1 and this aligns upstream with downstream.

Fixes that need to apply to downstream as well have been pushed as separate commits.

No plans to change the commit and break it out any further than just pushed as the intent of the PR is downstream alignment. Feel free to close this PR if that is a requirement and I'll create the numerous required piecemeal github issues to address the various changes this PR covers.

File: src/NzbDrone.Core/Download/DownloadClientProvider.cs [hunk 7]
```diff
@@ -111,7 +111,7 @@ private IEnumerable<IDownloadClient> FilterBlockedIndexers(IEnumerable<IDownload
             {
                 if (blockedClients.TryGetValue(client.Definition.Id, out var blockedClientStatus))
                 {
-                    _logger.Debug("Temporarily ignoring client {0} till {1} due to recent failures.", client.Definition.Name, blockedClientStatus.DisabledTill.Value.ToLocalTime());
+                    _logger.Debug("Temporarily ignoring client {0} until {1} due to recent failures.", client.Definition.Name, blockedClientStatus.DisabledTill.Value.ToLocalTime());
```
Comment (markus101) [inline src/NzbDrone.Core/Download/DownloadClientProvider.cs:114]: This doesn't need a public change log message

File: src/NzbDrone.Core/HealthCheck/Checks/ImportListRootFolderCheck.cs (hunk 4 above)
Comment (markus101) [reply src/NzbDrone.Core/HealthCheck/Checks/ImportListRootFolderCheck.cs]: Marking the conversation as resolved doesn't answer the question.

File: src/NzbDrone.Core/HealthCheck/Checks/IndexerSearchCheck.cs (hunk 6 above)
Comment (markus101) [reply src/NzbDrone.Core/HealthCheck/Checks/IndexerSearchCheck.cs:26]: This was not resolved.

File: src/NzbDrone.Core/HealthCheck/Checks/DownloadClientSortingCheck.cs (hunk 5 above)
Comment (markus101) [reply src/NzbDrone.Core/HealthCheck/Checks/DownloadClientSortingCheck.cs:2]: @mynameisbogdan did these in a separate commit, they can be removed.

**Review by markus101 — CHANGES_REQUESTED** [review]
> No plans to change the PR and break it out any further than what is just pushed as the intent of the PR is downstream alignment.

If the separate commits that need to go downstream can be separated then the original commit can also be broken up, your commit already calls out the separate things that are being done.

> Feel free to close this PR if that is a requirement and we can create the numerous required piecemeal github issues to address the various changes this PR covers.

We're not going to create a bunch of issues for minor discrepancies that can be fixed by rewriting the git history. https://stackoverflow.com/a/6217314 has a good explanation on how to do it. If this is something you're not comfortable doing then it'll need to wait for someone else to do it.

This PR will also need to be rebased after recent changes in develop.
